database/database.go

```diff
@@ -486,3 +489,75 @@ func (s *DatabaseService) DeleteExecutionPayloads(idFirst, idLast uint64) error
+	if err != nil {
+		return err
+	}
+	bidTrace := submitBlockRequest.Message
+	builderDemotionEntry := BuilderDemotionEntry{
+		SubmitBlockRequest: NewNullString(string(_submitBlockRequest)),
+
+		Slot:  bidTrace.Slot,
+		Epoch: bidTrace.Slot / uint64(common.SlotsPerEpoch),
+
+		BuilderPubkey:  bidTrace.BuilderPubkey.String(),
+		ProposerPubkey: bidTrace.ProposerPubkey.String(),
+
+		Value: bidTrace.Value.String(),
+
+		BlockHash: bidTrace.BlockHash.String(),
+	}
+
+	if signedBlindedBeaconBlock != nil {
+		_signedBlindedBeaconBlock, err := json.Marshal(signedBlindedBeaconBlock)
+		if err != nil {
+			return err
+		}
+		builderDemotionEntry.SignedBlindedBeaconBlock = NewNullString(string(_signedBlindedBeaconBlock))
+	}
+
+	if signedValidatorRegistration != nil {
+		_signedValidatorRegistration, err := json.Marshal(signedValidatorRegistration)
+		if err != nil {
+			return err
+		}
+		builderDemotionEntry.SignedValidatorRegistration = NewNullString(string(_signedValidatorRegistration))
+		builderDemotionEntry.FeeRecipient = signedValidatorRegistration.Message.FeeRecipient.String()
+		builderDemotionEntry.GasLimit = signedValidatorRegistration.Message.GasLimit
+	}
+
+	var query string
+	// If block_hash conflicts and we have a published block, fill in fields needed for the refund.
+	if signedBlindedBeaconBlock != nil && signedValidatorRegistration != nil {
+		query = `INSERT INTO ` + vars.TableBuilderDemotions + `
+			(submit_block_request, signed_blinded_beacon_block, signed_validator_registration, slot, epoch, builder_pubkey, proposer_pubkey, value, fee_recipient, gas_limit, block_hash) VALUES
+			(:submit_block_request, :signed_blinded_beacon_block, :signed_validator_registration, :slot, :epoch, :builder_pubkey, :proposer_pubkey, :value, :fee_recipient, :gas_limit, :block_hash)
+			ON CONFLICT (block_hash) DO UPDATE SET
+				signed_blinded_beacon_block = :signed_blinded_beacon_block,
+				signed_validator_registration = :signed_validator_registration,
+				fee_recipient = :fee_recipient,
+				gas_limit = :gas_limit;
+			`
+	} else {
+		// If the block_hash conflicts, then all the relevant data must be there already, so do nothing.
+		query = `INSERT INTO ` + vars.TableBuilderDemotions + `
+			(submit_block_request, signed_blinded_beacon_block, signed_validator_registration, slot, epoch, builder_pubkey, proposer_pubkey, value, fee_recipient, gas_limit, block_hash) VALUES
+			(:submit_block_request, :signed_blinded_beacon_block, :signed_validator_registration, :slot, :epoch, :builder_pubkey, :proposer_pubkey, :value, :fee_recipient, :gas_limit, :block_hash)
+			ON CONFLICT (block_hash) DO NOTHING`
+	}
+	_, err = s.DB.NamedExec(query, builderDemotionEntry)
+	return err
+}
+
+func (s *DatabaseService) GetBlockBuildersFromCollateralID(collateralID uint64) ([]*BlockBuilderEntry, error) {
+	query := `SELECT builder_pubkey FROM ` + vars.TableBlockBuilder + ` WHERE collateral_id=$1 ORDER BY id ASC;`
```

Comment:
Suggest removing `ORDER BY id ASC`. Shorter and faster query.

Reply:
done!